config/reader.go

```diff
@@ -49,29 +57,150 @@ func getAgentInfo() (*apppb.AgentInfo, error) {
 	}, nil
 }
 
-var viamDotDir = filepath.Join(os.Getenv("HOME"), ".viam")
+// FileReader responsible for converting the raw file into a Config type
+type FileReader struct {
+	// filepath is the filepath of the config
+	Filepath string
+	// config is the output of the reader and what gets manipulated
+	Config *Config
 
-func getCloudCacheFilePath(id string) string {
-	return filepath.Join(viamDotDir, fmt.Sprintf("cached_cloud_config_%s.json", id))
+	fileInBytes      []byte
+	localDataDir     string
+	localPackagesDir string // path.Join(viamDotDir, "packages")
+
+	// CurrentPackages stores a list of valid packages used to try and update placeholders in the
+	// entire config
+	CurrentPackages map[string]string
+}
+
+// copied from package-manager
+func HashName(f PackageConfig) string {
+	// replace / to avoid a directory path in the name. This will happen with `org/package` format.
+	// also makes the path valid by replacing all the dots with _
+	return fmt.Sprintf("%s-%s", strings.ReplaceAll(f.Package, "/", "-"), HashVersion(f.Version))
+}
+
+func HashVersion(version string) string {
+	// replaces all the . if they exist with _
+	return strings.ReplaceAll(version, ".", "_")
+}
+
+// there are 2 ways that we have configs, from rawfiles or from configs from the cloud
+// these methods do similar things, but in different orders
+
+func replacePlaceholders(bytes []byte, packageMap map[string]string) []byte {
+	// updates all the filepath placeholders with what we think they should be based on the packages
+	updatedBytes := placeholderRegexp.ReplaceAllFunc(bytes, func(b []byte) []byte {
+		updatedPath, ok := packageMap[verifyPlaceholder(string(b[:]))]
+		if !ok {
+			return b
```

Comment:
we currently don't provide a warning with packages so was following that convention, because when you then go to download it it will fail anyways. Will add one in though!

Reply:
The error that comes up after this is that the file does not exist at location: `${packges.x.y...}`. In this case, I can throw an error here that would stop the config from being processed and from the robot starting. Will add!